plutus-core/plutus-ir/src/PlutusIR/Transform/NewLetFloat.hs

```diff
@@ -0,0 +1,493 @@
+            -> c -- ^ term or bindings to float around
+            -> re c -- ^ the combined result
+    floatAt toPos floatFunc termOrBindings = do
+        herePos <- toPos <$> ask
+        -- is there something to be floated here?
+        case fTable M.!? herePos  of
+            -- nothing to float, just descend
+            Nothing -> pure termOrBindings
+            -- all the naked-lets to be floated here
+            Just letsNaked -> do
+                -- visit the rhs'es of these floated lets for any potential floatings as well
+                -- NOTE: we do not directly run `go(letsNaked)` because that would increase the depth,
+                -- and the floated lets are not anchors themselves; instead we run go on the floated-let bindings' subterms.
+                letsNaked' <- letsNaked & (traversed.letBs.traversed.bindingSubterms) go
+                -- apply the merging with the visit result. This is what floats them back to the pir.
+                pure $ floatFunc letsNaked' termOrBindings
+
+-- | Squeezes floatable lets' (naked lets) bindings next to the bindings of the "parent" unfloatable let, that floatable lets depends upon.
+-- See [Floating rhs-nested lets]
+squeezeBindings :: NE.NonEmpty (LetNaked tyname name uni fun a) -- ^ the floatable lets
+                -> NE.NonEmpty (Binding tyname name uni fun a) -- ^ the bindings of the unfloatable let-group we want to squeeze into
+                -> NE.NonEmpty (Binding tyname name uni fun a) -- ^ the result "larger" bindings
+squeezeBindings floatableNaked unfloatableBindings =
+    -- TODO: we lose the annotations of the naked lets, fix with semigroup.
+    concatNaked floatableNaked <> unfloatableBindings
+
+-- | Create a brand-new letrec to group the floated lets.
+makeNewLet :: NE.NonEmpty (LetNaked tyname name uni fun a) -> Term tyname name uni fun a -> Term tyname name uni fun a
+makeNewLet floatableNaked =
+    -- arbitrary: use the annotation of the first let of the floated lets as the annotation of the new let
+    -- TODO: use some semigroup annotation-joining instead
+    Let (NE.head floatableNaked^.letAnn)
+    Rec -- needs to be rec because we don't do dependency resolution at this pass, See Note [LetRec splitting pass]
+    $ concatNaked floatableNaked
+
+-- bring the bindings of every naked let together
+concatNaked :: NE.NonEmpty (LetNaked tyname name uni fun a)
+            -> NE.NonEmpty (Binding tyname name uni fun a)
+concatNaked = foldMap1 (^.letBs)
+
+
+-- | The compiler pass of the algorithm (comprised of 3 connected passes).
+floatTerm :: (PLC.ToBuiltinMeaning uni fun,
+            PLC.HasUnique tyname PLC.TypeUnique, PLC.HasUnique name PLC.TermUnique,
+            Ord tyname, Ord name
+            )
+          => Term tyname name uni fun a -> Term tyname name uni fun a
+floatTerm t =
+    mark t
+    & flip removeLets t
+    & uncurry floatBackLets
+
+-- CONSTANTS
```

Comment:
I feel slightly uneasy about arbitrarily chosen constraint.

I do not remember which versions of GHC we are targeting, but after 8.4 you can use `Maybe (Max Pos)` from `Data.Semigroup` to calculate maximum on `Pos` with a meaningful minimal element `Nothing`.

This way, you could calculate `maxPos` as

```haskell
type MaxPos = Maybe (Max Pos)

maxPos :: M.Map k Pos -> MaxPos
maxPos = foldMap (Just . Max)
```
so that you will never confuse the distinguished position for  top (`Nothing`) for a valid position.

The added bonus is that we are using `Foldable` instance for `M.Map`, so it could be more optimised that the instance for `List`.  I am not sure if it matters in this case, but one comment mentions switching to `UniqueMap`, which IIRC has `IntMap` under the hood, and it that case it might matter.

Finally, I am not sure how difficult it would be to update your algorithm to mean `Nothing` for top `Pos`...

Reply:
I agree this is a bit ugly, but I think it's okay for now. It would be nice to instead have a datatype for `Pos` with a `Top` alternative, but that would probably require fiddling with lots of pieces.